Node-sdk (#147)
bucketco/bucket-javascript-sdk: packages/node-sdk/README.md

// other custom properties.
//
// We recommend that only one global instance of `client` should be created
// to avoid multiple round-trips to our servers.
const client = new Client({ secretKey: "sec_prod_xxxxxxxxxxxxxxxxxxxxx" });

// Initialize the client's feature flag retrieval loop. This is not a mandatory
// call, but it is recommended in order to warm up the cache so that future
// calls to `getFlags()` will have cached flag definitions available.
//
// You can also supply your `fallbackFlags` to the `initialize()` method. These
// fallback flags are used in the situation when the server-side flags are not
// yet loaded or there are issues retrieving them.
await client.initialize();
```

At any point where the client needs to be used, we can configure it through
`withUsr()`, `withCompany()` and `withCompanyContext()` methods. Each of
these three methods returns a new instance of `Client` class which includes
the applied user, company or context:

```ts
// configure the client
const boundClient = client
  .withUser({ id: "john_doe" name: "John Doe" })
  .withCompany({ id: "acme_inc", name "Acme Inc."})

>   .withCustomContext({ additional: "context", number: [1,2,3] })

Reviewer: Do these chains result in a new immutable instance for each call, or does the first one create a new instance which is then mutated by the remaining calls?

Only could easily expect user, company and context to be available in server middlewares at different points, so it won't be possible to set up this chain once and for all in a single place. Rather these pieces of information will have to be added piecemeal. If a new immutable instance is created each time, the client reference will have to be re-assigned in each place. It might be relevant to show an example of this.

If I wanted to check a flag without any user, company or custom context, would I just access the global client instance? I'm thinking of when we presumably add always on/off rules per environment. These could be evaluated fine without any additional context other than the secret key

Author: We discussed and came up with changes.

```

See [types](./src/types.ts) for more details on the supported types.

You can then then use the `boundClient` to perform all the supported functionality:

```ts
// update the user with Bucket (await it optional)

> await boundClient.updateUser();

Reviewer: This example doesn't make a lot of sense. What would the effect of calling this function without any arguments be?

// update the associated company with Bucket

> await boundClient.updateCompany();

Reviewer: This example doesn't make a lot of sense. What would the effect of calling this function without any arguments be?

  refetchInterval?: number = 60000,
  // The interval to re-fetch feature flags (optional). Default
  // is 5 times the re-fetch interval.
  staleWarningInterval?: number = 5 * 60000,
  // The logger you can supply. By default no logging is performed.
  logger?: Logger;
  // The custom http client. By default the internal `fetchClient` is used.
  httpClient?: HttpClient = fetchCient,
}
```

### Zero PII

The Bucket SDK doesn't collect any metadata and HTTP IP addresses are _not_ being
stored. For tracking individual users, we recommend using something like database
ID as userId, as it's unique and doesn't include any PII (personal identifiable
information). If, however, you're using e.g. email address as userId, but prefer
not to send any PII to Bucket, you can hash the sensitive data before sending
it to Bucket:

```ts
import { sha256 } from 'crypto-hash';

client.withUser({ userId: await sha256("john_doe"), ... });
```

### Custom Attributes & Context

You can pass attributes as part of the object literal passed to the `withUser()`,
`withCompany()` and `trackFeatureUsage()` methods. Attributes cannot be nested
(multiple levels) and must be either strings, integers or booleans.

> Some attributes are used by Bucket.co and manged separately, such as:

Reviewer: ```suggestion
Some attributes are used by Bucket.co to improve the UI, and are recommended to provide for easier navigation:
```

- `name` or `$name` -- display name for `user`/`company`,
- `email` or `$email` -- the email of the user.

You can supply an additional `context` to these functions as well. The `context`
object contains additional data that Bucket uses to make some behavioural choices.

By default, sending `updateCompany`, `updateUser` and `trackFeatureUsage` calls
automatically update the given user/company `Last seen` property. You can control
if `Last seen` should be updated when the events are sent by setting
`context.active = false` to avoid updating last seen. This is often useful if you
have a background job that goes through a set of companies just to update their.

Example:

```ts
// create a new client initialized with an user and a company.
const boundClient = client
  .withUser({
    userId: "188762",
    attributes: { name: "John O." },
    context: { active: true },
  })
  .withCompany({
    companyId: "0083663",
    attributes: { name: "My SaaS Inc." },
    context: { active: false },
  });

// send updates to Bucket.co with the new details.
await boundClient.updateCompany();

> await boundClient.updateUser();

Reviewer: This seems like an unintuitive way of triggering the underlying user and company tracking events to bucket. You might well want to register different things in bucket than you'd want to provide as a feature flagging context.

Especially the `active: true/false` seems like something that is not relevant to feature flag checks at all, but it only relevant to updating entities in bucket. So I would expect the `active`-meta to be a part of the function that triggers these bucket HTTP API calls

# Bucket Node.js SDK

Node.js, JavaScriptS/Typescript feature flag and tracking client for [Bucket.co](https://bucket.co).

## Installation

Install using `yarn` or `npm` with:

> `yarn add -s @bucketco/node-sdk` or `npm install -s @bucketco/node-sdk`.

> Other languages than JavaScript/Typescript are currently not supported by an SDK.
> You can [use the HTTP API directly](https://docs.bucket.co/reference/http-tracking-api)

Reviewer 2: ```suggestion
Other supported languages/frameworks are in the [Supported languages](https://docs.bucket.co/quickstart/supported-languages) documentation pages.

You can also [use the HTTP API directly](https://docs.bucket.co/reference/http-tracking-api)
```

## Basic usage

> Before the library can be used, you need to obtain a secret key from
> [Environment setting view](https://app.bucket.co/envs/{environment}/settings/app-environments)
> in **Bucket.co**.
> 
> > [!CAUTION]
> > Secret keys should only be used by the `node-sdk` and not any other
> > SDKs provided by Bucket. Secret keys offer the users the ability to obtain
> > information that is often sensitive and thus should not be used in any
> > front-end application.

Reviewer 2: ```suggestion
To get started you need to obtain a secret key from
[Environment setting view](https://app.bucket.co/envs/{environment}/settings/app-environments)
in **Bucket.co**.

> [!CAUTION]
> Secret keys are meant for use in server side SDKs only.
> Secret keys offer the users the ability to obtain
> information that is often sensitive and thus should not be used in
> client-side applications.
```

```ts
import { BucketClient } from "@bucketco/node-sdk";

// Create a new instance of the client with the secret key. Additional options
// are available, such as supplying a logging sink, a custom HTTP client and
// other custom properties.
//
// We recommend that only one global instance of `client` should be created
// to avoid multiple round-trips to our servers.
const client = new BucketClient({
  secretKey: "sec_prod_xxxxxxxxxxxxxxxxxxxxx",
});

> // Initialize the client's feature flag retrieval loop. You must call this method
> // prior to any calls to `getFlags()`, otherwise an empty object will be returned.
> //
> // You can also supply your `fallbackFlags` to the `initialize()` method. These
> // fallback flags are used in the situation when the server-side flags are not
> // yet loaded or there are issues retrieving them.
> await client.initialize();

Reviewer 2: ```suggestion
// Initialize the client and begin fetching flag definitions.
// You must call this method prior to any calls to `getFlags()`,
// otherwise an empty object will be returned.
//
// You can also supply your `fallbackFlags` to the `initialize()` method. These
// fallback flags are used in the situation when the server-side flags are not
// yet loaded or there are issues retrieving them.
// See "Initialization Options" below for more information.
await client.initialize();
```

```

At any point where the client needs to be used, we can configure it through
`withUser()`, `withCompany()` and `withCompanyContext()` methods. Each of
these three methods returns a new instance of `Client` class which includes
the applied user, company or context:

```ts
// configure the client
const boundClient = client
  .withUser("john_doe", { attributes: { name: "John Doe" } })
  .withCompany("acme_inc", { attributes: { name "Acme Inc."} })
  .withCustomContext({ additional: "context", number: [1,2,3] })

const fallbackFlags = { can_see_new_reports: { value: true } }

// actively wait for the flags to be loaded

> await boundClient.initialize(fallbackFlags)

Reviewer 2: Hm, I think fallbackFlags should be an argument to the `Client` constructor instead of to `initialize`?  Feels like it fits into the `Client` constructor with the other initialization arguments.

Reviewer 2: i think the order here is a bit funky. We just showed how to bind the client and here comes the initialization example to show how to use the fallbackFlags. Let's make `fallbackFlags` part of the Client constructor and remove this example. I want the getting started examples to be as simple as possible so the `Client` creation example you have above that just takes the secretKey is good. Let's just explain the fallbackFlag in "Initialization Options" below.

Author: Hm, so we won't be able to switch over to new fallback flags when we want? We would only be able to create a new instance of the client when we need new fallback flags.

Reviewer 2: Was the idea that you can call initialize again if you want to change the fallback flags? I’d expect you’d be unable to call initialize more than once so that would be a surprising recommendation to me. I think it’s fine to require folks to create a new client if they want to switch the fallback flags.

is used to track feature usage across your application.

The following example shows how to register a new user, and associate it with a company:

```ts
// registers the user with Bucket using the provided unique ID, and
// providing a set of custom attributes (can be anything)
client.trackUser("your_user_id", {
  attributes: { longTimeUser: true, payingCustomer: false },
});

// register the user as being part of a given company
client.trackCompany("company_id", { userId: "your_user_id" });
```

If one want to simply update a company's attributes on Bucket side,
one calls `trackCompany` without supplying an user ID:

```ts
// either creates a new company on Bucket or updates an existing
// one by supplying custom attributes
client.trackcompany("updated_company_id", {
  attributes: {
    status: "active",
    plan: "trial",
  },
});

// if a company is not active, and one needs to make sure its
// "Last Seen" status does not get updated, one can supply
// an additional meta argument at the end:
client.trackcompany("updated_company_id", {
  attributes: { status: "active", plan: "trial" },
  meta: { active: false },
});
```

To generate feature tracking `event`s:

```ts

> // this simply sends an event to Bucket, not associated with any company or user.

Author: see comment below

client.trackFeatureUsage("some_feature_name");

// to specify to which user/company this event belongs one can do
client.trackFeatureUsage("some_feature_name", {
  companyId: "acme_inc",
  userId: "73668762",
});
```

### Initialization Options

Supply these to the `constructor` of the `Client` class:

```ts
{
  // The secret key used to authenticate with the Bucket API.
  secretKey: string,
  // The host to send requests to (optional).
  host?: string = "https://tracking.bucket.co",
  // The interval to re-fetch feature flag definitions (optional).
  // Default is 60 seconds.
  refetchInterval?: number = 60000,
  // The interval to re-fetch feature flags (optional). Default
  // is 5 times the re-fetch interval.
  staleWarningInterval?: number = 5 * 60000,
  // The logger you can supply. By default no logging is performed.
  logger?: Logger,
  // The custom http client. By default the internal `fetchClient` is used.

>   httpClient?: HttpClient = fetchCient,

Reviewer 2: missing `fallbackFlags`

client.withUser({ userId: await sha256("john_doe"), ... });
```

### Custom Attributes & Context

You can pass attributes as part of the object literal passed to the `withUser()`,
`withCompany()` and `trackFeatureUsage()` methods. Attributes cannot be nested
(multiple levels) and must be either strings, integers or booleans.

Some attributes are used by Bucket.co to improve the UI, and are recommended
to provide for easier navigation:

- `name` or `$name` -- display name for `user`/`company`,
- `email` or `$email` -- the email of the user.

You can supply an additional `context` to these functions as well. The `context`
object contains additional data that Bucket uses to make some behavioural choices.

By default, `trackUser`, `trackCompany` and `trackFeatureUsage` calls
automatically update the given user/company `Last seen` property on Bucket side.
You can control if `Last seen` should be updated when the events are sent by setting
`context.active = false` to avoid updating last seen. This is often useful if you
have a background job that goes through a set of companies just to update their.

Example:

```ts
// create a new client initialized with an user and a company.

> client.trackUser("188762", {

Author: That was the initial implementation but after discussions with @Munter we settled on a manual approach. 

Otherwise I'd have to introduce 3 "bound classes": `bound-to-user`, `bound-to-company` and `bound-to-both`. Not a biggie but annoying enough.

> client.trackUser("188762", {

Reviewer 2: What’s the point of “withUser” if you must still pass userId to track?

> client.trackUser("188762", {

Reviewer 2: apologies, I'm mixing things up. `trackFeatureUsage` doesn't take a `userId`. So nevermind 👍

> client.trackUser("188762", {

Author: No, you were correct, we require user/company explicitly supplied for any `trackXXXX` method. I will make the changes to use bound clients if that's the decision. I wish @Munter was here since he had the opposite view. Would have liked to have all opinions on the table.

  secretKey: "sec_prod_xxxxxxxxxxxxxxxxxxxxx",
  fallbackFlags: { can_see_new_reports: true },
});

// Initialize the client and begin fetching flag definitions.
// You must call this method prior to any calls to `getFlags()`,
// otherwise an empty object will be returned.
//
// You can also supply your `fallbackFlags` to the `initialize()` method. These
// fallback flags are used in the situation when the server-side flags are not
// yet loaded or there are issues retrieving them.
// See "Initialization Options" below for more information.
await client.initialize();
```

At any point where the client needs to be used, we can configure it through
`withUser()`, `withCompany()` and `withOtherContext()` methods. Each of
these three methods returns a new instance of `Client` class which includes
the applied user, company or context:

```ts
// configure the client
const boundClient = client
  .withUser("john_doe", { attributes: { name: "John Doe" } })
  .withCompany("acme_inc", { attributes: { name "Acme Inc."} })
  .withOtherContext({ additional: "context", number: [1,2,3] })

const fallbackFlags = { can_see_new_reports: true  }

// actively wait for the flags to be loaded

> await boundClient.initialize()

Author: It doesn't really matter when it's called, though the earlier - the better. Any subsequent calls will effectively refresh the flags. But let me correct the docs

```

Once the client is initialized, one can obtain the evaluated flags:

```ts
// get the current flags (uses company, user and custom context to evaluate the flags).
const flags = boundClient.getFlags();

if (flags.can_see_new_reports) {
  // this is your flag-protected code ...
  // send an event when the feature is used:
  boundClient.trackFeatureUsage("new_reports_used", {
    attributes: {
      some: "attribute",
    },
  });
}
```

## Tracking

Tracking allows sending `user`, `company` and `event` messages to Bucket.

> `user` events can be used to register or update an user's attributes

Reviewer 2: ```suggestion
`user` events can be used to register or update a user's attributes
```

with Bucket. `company` allows the same, and additionally, allows
associating an user with a company on the Bucket side. Finally, `event`
is used to track feature usage across your application.

The following example shows how to register a new user, and associate it with a company:

```ts
// registers the user with Bucket using the provided unique ID, and
// providing a set of custom attributes (can be anything)
const boundClient = client
  .withUser("your_user_id", {
    attributes: { longTimeUser: true, payingCustomer: false },
  })
  .withCompany("company_id");

// track the user (send a `user` event to Bucket).
await boundClient.trackUser();

// register the user as being part of a given company

> boundClient.trackCompany();

Reviewer 2: Maybe you and @Munter were already back and forth on this, but did you consider removing the `trackCompany` / `trackUser` methods and instead only relying on `withCompany` / `withUser` ? This is a feature flagging SDK and whenever you do `withCompany` and give it some attributes and then go on to evaluate some flags, those attributes need to be transmitted to our servers in order for the views/tables there to be accurate (who has access to what). So, I think it makes sense for `withCompany` to automatically call `trackCompany` at the least. And then we should ask ourselves if `trackCompany` and `trackUser` is even useful as separate methods.

```

If one needs to simply update a company's attributes on Bucket side,
one calls `updateCompany` without supplying a user ID:

```ts
// either creates a new company on Bucket or updates an existing
// one by supplying custom attributes

> client.withCompany("updated_company_id").updateCompany({

Reviewer 2: ```suggestion
client.withCompany("company_id").updateCompany({
```
